### ISSUE TITLE
Integrating with Salesforce as an OpenID Connect external identity provider

### ISSUE THREAD

**Issue by duribe76** [IdentityServer/IdentityServer3#2800]
- [ X ] I read and understood how to enable [logging](https://identityserver.github.io/Documentation/docsv2/configuration/logging.html)

### Question / Issue

We have some users in our application that are already part of a Salesforce instance and we want them to be able to use their existing credentials, while other users will be using the internal repository (with MembershipReboot). Salesforce now offers an OpenID connect identity provider (https://developer.salesforce.com/page/Inside_OpenID_Connect_on_Force.com) and I was trying to set it up as an external provider for IdP, but I'm not having much luck with it. Below is the relevant configuration adding the external providers from my OWIN startup and log. I am getting to Salesforce, but when it goes back to IdP, it doesn't complete the process, it just shows the welcome page with code and state query string parameters.

```
public static void ConfigureIdentityProviders(IAppBuilder app, string signInAsType)
        {
            app.UseOpenIdConnectAuthentication(
                new OpenIdConnectAuthenticationOptions
                {
                    AuthenticationType = "Salesforce",
                    Authority = "https://login.salesforce.com/",
                    Caption = "Salesforce",
                    ClientId = "3MVG94DzwlYDSHS7gUmSAfJutM0Q3K9wbcMUkdOTk_lPEE6Zc8RGbxEXc9EJk3KiYyOyIZAdUCqH7gkJmwded",
                    ClientSecret = "6079254137949426961",
                    RedirectUri = "https://identity.dev.local",
                    ResponseType = "code",
                    Scope = "openid id",
                    SignInAsAuthenticationType = signInAsType
                }
            );
        }
```

### Relevant parts of the log file, with sensitive data removed

```
2016-04-19 13:55:13,035 [61] INFO  IdentityServer3.Core.ResponseHandling.AuthorizeInteractionResponseGenerator - User is not authenticated. Redirecting to login.
2016-04-19 13:55:13,040 [61] INFO  IdentityServer3.Core.Endpoints.AuthorizeEndpointController - End authorize request
2016-04-19 13:55:13,040 [61] INFO  IdentityServer3.Core.Results.LoginResult - Redirecting to login page
2016-04-19 13:55:13,041 [61] DEBUG IdentityServer3.Core.Configuration.Hosting.MessageCookie`1 - Protecting message: {"ReturnUrl":"https://identity.dev.local/connect/authorize?client_id=...&redirect_uri=...&response_mode=form_post&response_type=code%20id_token&scope=openid%20profile%20email%20address%20phone%20roles%20accounts%20offline_access&state=...&nonce=...","ClientId":"...","AcrValues":[],"Created":635966853126883680}
2016-04-19 13:55:13,048 [62] INFO  IdentityServer3.Core.Endpoints.AuthenticationController - Login page requested
2016-04-19 13:55:13,048 [62] DEBUG IdentityServer3.Core.Endpoints.AuthenticationController - signin message passed to login: {
  "ReturnUrl": "https://identity.dev.local/connect/authorize?client_id=...&redirect_uri=...&response_mode=form_post&response_type=code%20id_token&scope=openid%20profile%20email%20address%20phone%20roles%20accounts%20offline_access&state=...&nonce=...",
  "ClientId": "...",
  "IdP": null,
  "Tenant": null,
  "LoginHint": null,
  "DisplayMode": null,
  "UiLocales": null,
  "AcrValues": [],
  "Created": 635966853126883680
}
2016-04-19 13:55:13,058 [62] INFO  IdentityServer3.Core.Endpoints.AuthenticationController - rendering login page
2016-04-19 13:55:19,228 [54] INFO  IdentityServer3.Core.Endpoints.AuthenticationController - External login requested for provider: Salesforce
2016-04-19 13:55:19,235 [54] INFO  IdentityServer3.Core.Endpoints.AuthenticationController - Triggering challenge for external identity provider
2016-04-19 13:55:19,787 [59] INFO  IdentityServer3.Core.Endpoints.WelcomeController - Welcome page requested - rendering
```

**Comment by leastprivilege** [comment]
You would first need to figure out if the Microsoft middleware is compatible with SalesForce at all. That's not an IdentityServer question.

e.g. - IIRC the MS MW does not support `code` response type.

**Comment by totpero** [comment]
@duribe76 you can request implementation for this external identity provider here: https://github.com/tparnell8/OwinOAuthProviders

**Comment by duribe76** [comment]
@leastprivilege, I think you probably hit the point. I first tried to use the Hybrid flow (ResponseType = "code id_token") and the Salesforce Open ID provider didn't accept it. Since you were already familiar with this OWIN middleware issue, Is there any ticket on that project that we can follow to know if and when the authorization code flow might be supported? Since Salesforce also supports SAML2p, so we may give that a shot.

**Comment by leastprivilege** [comment]
There are no investments from Microsoft in Katana anymore.

Maybe try the 3rd party salesforce middleware instead.

**Comment by duribe76** [comment]
Yes, I noticed that after doing more research and finding the OWIN/Katana ticket for implementing the authentication code flow. However, the good news is that it seems that this is already addressed in the ASP.NET 5 version of the middleware, so we will probably consider moving our IdentityServer instance to that once it is out of beta and GA.

In the meantime, I was able to successfully integrate with Salesforce as an external IdP by enabling it as such (with SAML2) and using Kentor AuthServices middleware.

Thanks for all your prompt responses!
